sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/policy/RetryPolicy.java

```diff
@@ -86,31 +127,66 @@ public RetryPolicy(RetryStrategy retryStrategy, String retryAfterHeader, ChronoU
+    public static Duration getWellKnownRetryDelay(Headers responseHeaders, int tryCount, RetryStrategy retryStrategy,
+                                                  Supplier<OffsetDateTime> nowSupplier) {
+        Duration retryDelay = ImplUtils.getRetryAfterFromHeaders(responseHeaders, nowSupplier);
+        if (retryDelay != null) {
+            return retryDelay;
+        }
+
+        // None of the well-known headers have been found, return the default delay duration.
+        return retryStrategy.calculateRetryDelay(tryCount);
     }
 
     /**
-     * Creates a {@link RetryPolicy} with the provided {@link RetryOptions}.
-     *
-     * @param retryOptions The {@link RetryOptions} used to configure this {@link RetryPolicy}.
-     *
-     * @throws NullPointerException If {@code retryOptions} is null.
+     * The interface for determining the retry strategy of clients.
      */
-    public RetryPolicy(RetryOptions retryOptions) {
-        this(ImplUtils.getRetryStrategyFromOptions(retryOptions), null, null);
-    }
+    @FunctionalInterface
+    public interface RetryStrategy {
 
-    @Override
-    public HttpResponse process(HttpRequest httpRequest, HttpPipelineNextPolicy next) {
-        return attempt(httpRequest, next, 0, null);
+        /**
+         * Computes the delay between each retry.
+         *
+         * @param retryAttempts The number of retry attempts completed so far.
+         * @return The delay duration before the next retry.
+         */
+        Duration calculateRetryDelay(int retryAttempts);
+
+        default Duration calculateExponentialRetryDelay(int retryAttempts, Duration baseDelay, Duration maxDelay) {
+            long baseDelayNanos = baseDelay.toNanos();
+            long maxDelayNanos = maxDelay.toNanos();
+            // Introduce a small amount of jitter to base delay
+            long delayWithJitterInNanos = ThreadLocalRandom.current()
+                .nextLong((long) (baseDelayNanos * (1 - JITTER_FACTOR)), (long) (baseDelayNanos * (1 + JITTER_FACTOR)));
+            return Duration.ofNanos(Math.min((1L << retryAttempts) * delayWithJitterInNanos, maxDelayNanos));
+        }
+
+        default boolean shouldRetry(HttpResponse httpResponse) {
+            int code = httpResponse.getStatusCode();
+            return (code == HttpURLConnection.HTTP_CLIENT_TIMEOUT
+                || (code >= HttpURLConnection.HTTP_INTERNAL_ERROR
+                && code != HttpURLConnection.HTTP_NOT_IMPLEMENTED
+                && code != HttpURLConnection.HTTP_VERSION));
+        }
+
+        /**
+         * This method is consulted to determine if a retry attempt should be made for the given {@link Throwable}
+         * propagated when the request failed to send.
+         *
+         * @param exception The {@link Throwable} thrown during the previous attempt.
+         * @return Whether a retry should be attempted.
+         */
+        default boolean shouldRetryException(Exception exception) {
```

Comment:
@JonathanGiles @srnagar this is a good time to right our regrets here, do we want to restrict the default retried exceptions here?